New addon: Fuzzy Search with PostgreSQL pg_trgm extension

christophlsa opened the pull request:
This addon provides the ability to create GIN or GiST indexes of char and text fields and also to use the search operator `%` in search domains. This is a first basic implementation which should be extended later.
## Requirements

First you need to have the `pg_trgm` extension available. In debian based distribution you have to install the postgresql-contrib module. Then you have to either install the `pg_trgm`extension to your database or you have to give your postgresql user the superadmin right (then the extension will be installed by this module).
## Explanation

If you installed the odoo module you can define in Settings -> Database Structure -> Trigam Index the indexes. There the index name will automatically created.

For example you can create an index for the `name` field of `res.partner`. Then in a search you can use `self.env['res.partner'].search([('name', '%', 'Jon Miller)])`. In this Example it can find existing names like `John Miller` or `John Mill`.

Which strings can be found depends on the limit which is set (default: 0.3). Currently you have to set the limit by executing SQL: `self.env.cr.execute("SELECT set_limit(0.2);")`.

Also you can use the `similarity(column, 'text')` function in the `order` parameter to order by the similarity. This is just a basic implementation which doesn't contains validations and has to start with this function. For example you can define this like `"similarity(%s.name, 'John Mil') DESC" % self.env['res.partner']._table`.

For further questions read the Documentation of the [pg_trgm](https://www.postgresql.org/docs/current/static/pgtrgm.html) module.
## Tasks/Goals

**For this PR:**
- [x] Set permissions/groups
- [x] Add possibility to order by similarity
- [x] Add Readme etc
- [x] German Translations
- [x] Add unittests

**Later:**
- Modify the general search parts (e.g. in tree view or many2one fields)

Commit:
[ADD] First implementation of base_trgm_search for creating indexes and allowing the sql parameter %.

Commit:
[IMP] Allow to order by similarity.

lmignon commented on base_trgm_search/models/trgm_index.py:
```diff
@@ -0,0 +1,224 @@
+def patch_generate_order_by(method):
+    def decorate_generate_order_by(self, order_spec, query):
+        if order_spec and order_spec.startswith('similarity('):
+            from pprint import pprint
```
remove pprint lines

christophlsa replied:
Thanks. I think I committed to fast ^^

dreispt commented on base_trgm_search/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
+from . import models
```
I suggest to name the module as `base_search_similarity`.

StefanRijnhart replied:
base_search_fuzzy?

dreispt replied:
@StefanRijnhart :laughing: Yes. That sounds good too.

christophlsa replied:
I renamed it :)

dreispt commented on base_trgm_search/models/trgm_index.py:
```diff
@@ -0,0 +1,224 @@
+                                'extension pg_trgm in your database or you '
+                                'have to be the superuser.')
+        else:
+            return True
```
It would be nice to have an INFO message informing that trigram search is available.

christophlsa replied:
I added an info to the auto_init function.

dreispt commented on base_trgm_search/models/trgm_index.py:
```diff
@@ -0,0 +1,224 @@
+    return decorate_leaf_to_sql
+
+expression.expression._expression__leaf_to_sql = patch_leaf_trgm(
+    expression.expression._expression__leaf_to_sql)
```
I'm not sure if this patching will work.
Make sure you add tests.

StefanRijnhart replied:
Please perform patching in register_hook to prevent patching when the module is not installed on any database. And tests, yes!

christophlsa replied:
I moved the stuff to the register_hook. And yes this patching works - if not I wouldn't publish here. :D. But I created now unit tests to test the overwriting/patching and the functionality.

dreispt commented:
The initial approach looks good.
Take care with the technique used to patch the core, it can be tricky, so unit tests will be your friend.

Commit:
[IMP] Renamed module base_trgm_search to base_search_fuzzy, added Unit tests, added translations, added access permissions, moved the monkey patching to method _register_hook of ir.model and fixed _auto_init, added README, cleaned up some aprts

christophlsa commented:
- Renamed module base_trgm_search to base_search_fuzzy,
- added Unit tests
- added german translations
- added access permissions
  moved the monkey patching to method _register_hook of ir.model
- fixed _auto_init (wrong parameter)
- added README
- cleaned up some parts

dreispt commented:
Is it ready for review?

christophlsa commented:
for me yes :)

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+This addon provides the ability to create GIN or GiST indexes of char and text
+fields and also to use the search operator `%` in search domains. Currently
+this module doesn't change the backend search or something else. It provides
```
s/something/anything

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+This addon provides the ability to create GIN or GiST indexes of char and text
+fields and also to use the search operator `%` in search domains. Currently
+this module doesn't change the backend search or something else. It provides
+only the possibilty to do a fuzzy search for external addons.
```
possibility to perform

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+Installation
+============
+
+First you need to have the ``pg_trgm`` extension available. In debian based
```
please use enumerated list

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+Configuration
+=============
+
+If you installed the odoo module you can define ``GIN`` and ``GiST`` indexes for
```
Please use enumerated list

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+Usage
+=====
+
+For example you can create an index for the `name` field of `res.partner`. Then
```
Example: 
#.you can create an index for the `name` field of model `res.partner`. 
#.in the search function you can use

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+``self.env['res.partner'].search([('name', '%', 'Jon Miller)])``
+
+In this Example it can find existing names like `John Miller` or `John Mill`.
```
#.in this example the function will return positive result for  `John Miller` or `John Mill`.

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+In this Example it can find existing names like `John Miller` or `John Mill`.
+
+Which strings can be found depends on the limit which is set (default: 0.3).
```
#.you can tweak the number of string to be returned by adjusting the set limit (default: 0.3).
NB: Currently you have to set the limit by executing the following SQL statment:

elicoidal commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+``self.env.cr.execute("SELECT set_limit(0.2);")``
+
+Also you can use the ``similarity(column, 'text')`` function in the ``order``
```
Another interesting feature is the use of `similarity(column, 'text')` function in the `order`
parameter to order by similarity. This module just contains a basic implementation which
doesn't perform validations and has to start with this function. For example
you can define the function as followed:

elicoidal commented on base_search_fuzzy/models/ir_model.py:
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
```
please use short header

christophlsa replied:
done

elicoidal commented:
Thanks for the contribution!

moylop260 commented:
Please, Could you review our beta checks https://travis-ci.org/OCA/server-tools/jobs/139978373#L424-L430?

``` bash
************* Module base_search_fuzzy.models.trgm_index
base_search_fuzzy/models/trgm_index.py:122: [C8107(translation-required), TrgmIndex.create_index] String parameter of raise "UserError" requires translation. Use _('The pg_trgm extension does not exists or cannot be installed.')
```

Ignore

``` bash
************* Module base_search_fuzzy.models.ir_model
base_search_fuzzy/models/ir_model.py:24: [W0402(deprecated-module), ] Uses of a deprecated module 'openerp.osv'
************* Module base_search_fuzzy.tests.test_query_generation
base_search_fuzzy/tests/test_query_generation.py:3: [W0402(deprecated-module), ] Uses of a deprecated module 'openerp.osv'
```

StefanRijnhart commented:
@moylop260 FYI two of those are the dreaded false positive for imports from openerp.osv again. https://github.com/OCA/maintainer-quality-tools/issues/306. What is the recommended solution for that?

moylop260 commented:
@StefanRijnhart 
You are right I'll review this check

moylop260 commented:
👍 
(Just add _translate method in raises)
Good job! thanks

Commit:
[IMP] base_search_fuzzy: Updated README, translated UserError Exception, Added one more test

Commit:
[IMP] base_search_fuzzy: Updated license header

christophlsa commented:
Thanks for the comments. I updated all. :)

Commit:
[FIX] Fixed missing instance variable after cleanup

Commit:
base_search_fuzzy: [FIX] Check if functions are already decorated to avoid decorating the function many times whcih leads to recursion errors

christophlsa commented:
I added a fix to check if the functions which I overwrite are already decorated by the new decorator. We had problems that these functions got this decorator much times (in particular in the cron worker) and a maximum recursion depth error was thrown.

I also rebased the branch.

gurneyalex commented:
travis red build should be fixed by #507

gurneyalex commented on base_search_fuzzy/README.rst:
```diff
@@ -0,0 +1,104 @@
+#. The PostgreSQL extension ``pg_trgm`` should be available. In debian based
+   distribution you have to install the `postgresql-contrib` module.
+#. Install the ``pg_trgm`` extension to your database or give your postgresql
+   user the ``SUEPRUSER`` right (this allows the odoo module to install the
```
s/SUEPRUSER/SUPERUSER/

gurneyalex commented:
one small typo, but :+1:

dreispt committed:
Fix typo

dreispt commented:
PR https://github.com/blooparksystems/server-tools/pull/1

dreispt commented:
TravisCI is red 🚧

christophlsa committed:
Merge pull request #1 from dreispt/patch-1

Fix typo

christophlsa commented:
After rerunning the checks (and merging the Readme fix) all are green.
